### PR TITLE
[spv-in] Support atomics in fields of global structs

### PR DESCRIPTION
**Pull request by schell**
**Connections**

* https://github.com/gfx-rs/wgpu/issues/4489
* https://github.com/gfx-rs/wgpu/issues/6683

**Description**

Adds support for using atomic operations on specific fields of a global struct. 

**Testing**

These changes add a snapshot test for the SPIR-V frontend of a shader that uses an atomic op on one of three fields of a global struct. As expected, this shader fails to validate on `trunk`, resulting in `AtomicUpgradeError(MultiMemberStruct)`. 

These changes fix the atomic upgrade machinery to properly upgrade the type of only the specific field that the atomic operation is performed on. It does this by recording struct field access when the global variable is discovered and then unrolling those accesses later, during type upgrade.

**Checklist**

- [x] Rebase after #6692 
- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

**Comment by jimblandy** [comment]
@schell I pushed a commit with some doc fixes. Do they look okay?

**Comment by schell** [comment]
@jimblandy that looks good to me 👍 !

File: naga/src/front/atomic_upgrade.rs [hunk 1]
```diff
@@ -97,64 +101,72 @@ impl UpgradeState<'_> {
...
-                let binding = binding.clone();
-                let upgraded_member_type = self.upgrade_type(ty)?;
+                let index = field_indices.pop().ok_or(Error::UnexpectedEndOfIndices)?;
+                let mut new_members = vec![];
+                for (i, mut member) in members.clone().into_iter().enumerate() {
+                    if i == index as usize {
+                        member.ty = self.upgrade_type(member.ty, field_indices)?;
+                    }
+                    new_members.push(member);
+                }
```
Comment (jimblandy) [inline naga/src/front/atomic_upgrade.rs]: This seems much more complex than necessary. Isn't this equivalent?

```suggestion
                let index = field_indices.pop().ok_or(Error::UnexpectedEndOfIndices)? as usize;
                let mut new_members = members.clone();
                new_members[index].ty = self.upgrade_type(new_members[index].ty, field_indices)?;
```

**Comment by jimblandy** [comment]
I'm a little concerned that, if the input has a struct with N fields that are accessed as atomics, this will create N new copies of the struct, as we perform each upgrade. It feels like there's more to worry about here, although I can't put my finger on it this instant.

I'll come back to this in a bit.

**Comment by jimblandy** [comment]
@schell I was concerned about the fact that the algorithm might visit `Struct` types many types, once per each field accessed atomically, and per global via which it was reached.

I just pushed a commit as a "big suggestion" that ensures we visit types only once, and that we have all the information we need when we get there. Could you take a look and see what you think?

**Comment by jimblandy** [comment]
@teoxoy My last "big suggestion" commit here is large enough that I was wondering if someone else should look it over. What do you think?

**Comment by schell** [comment]
I think these changes look great! It would be good to get a snapshot test of a shader that uses atomics on multiple fields to ensure these are correct, which I'll work on, but I don't want to hold up this merging.

**Comment by teoxoy** [comment]
The changes in the last commit look good to me as well.

**Review by jimblandy — APPROVED** [review]
Okay, this all looks good to me, and since my changes weren't trivial, they have been reviewed by @teoxoy. So I think we're good to go.

**Comment by schell** [comment]
🚀

**Pull request merged**